src/QsCompiler/Core/Transformations/SyntaxTreeTransformation.fs

```diff
@@ -48,270 +45,218 @@ type SyntaxTreeTransformation<'T> private (state: 'T, options: TransformationOpt
 
     member this.SharedState = state
 
-    new(state: 'T, options: TransformationOptions) as this =
-        SyntaxTreeTransformation<'T>(state, options, "_internal_")
-        then
-            this.Types <- new TypeTransformation<'T>(this, options)
-            this.ExpressionKinds <- new ExpressionKindTransformation<'T>(this, options)
-            this.Expressions <- new ExpressionTransformation<'T>(this, options)
-            this.StatementKinds <- new StatementKindTransformation<'T>(this, options)
-            this.Statements <- new StatementTransformation<'T>(this, options)
-            this.Namespaces <- new NamespaceTransformation<'T>(this, options)
-
-    new(state: 'T) = new SyntaxTreeTransformation<'T>(state, TransformationOptions.Default)
+    new(state) = SyntaxTreeTransformation(state, TransformationOptions.Default)
```

Comment:
Does this need to be `SyntaxTreeTransformation<'T>`?

Reply:
The `'T` is inferred.

Reply:
Since we have another class that has the same name but is not generic, should we leave the `<T>` here just for code legibility, so that people don't get confused and think it is referring to the non-generic type, as I just did?

Reply:
In this case, it can't refer to the non-generic type, because this constructor has to call one of the alternate constructors of the same class. Also, the non-generic `SyntaxTreeTransformation` constructor doesn't take a `state` argument. I understand that there's some ambiguity here due to the name `SyntaxTreeTransformation` being overloaded, but this situation feels similar to other cases of type inference, where I prefer to let F# infer as much as it can and use the language tooling for more detailed type information.